engine/jobmaster/dm/ddl_coordinator.go

```diff
@@ -0,0 +1,839 @@
+}
+
+// handleDropTable handles drop table ddl.
+// remove source table from shard group.
+// mark shard group as deleted if all source tables are deleted.
+func (g *shardGroup) handleDropTable(ctx context.Context, item *metadata.DDLItem) bool {
+	_, ok := g.normalTables[item.SourceTable]
+	if !ok {
+		log.L().Warn("drop table does not exist", zap.Any("source table", item.SourceTable))
+		return false
+	}
+
+	delete(g.normalTables, item.SourceTable)
+	delete(g.conflictTables, item.SourceTable)
+	// nolint:errcheck
+	g.droppedColumnsStore.DelDroppedColumnForTable(ctx, item.SourceTable)
+	if len(g.normalTables) == 0 {
+		g.deleted = true
+		return true
+	}
+	return false
+}
+
+// handleDDLs handles ddl.
+func (g *shardGroup) handleDDLs(ctx context.Context, item *metadata.DDLItem) (newDDLs []string, conflictStage optimism.ConflictStage, err error) {
+	stmt, ok := g.normalTables[item.SourceTable]
+	if !ok || stmt == "" {
+		log.L().Warn("table does not exist", zap.Any("source table", item.SourceTable))
+		g.normalTables[item.SourceTable] = item.Tables[0]
+	}
+
+	dropCols := make([]string, 0, len(item.DDLs))
+
+	// handle ddls one by one
+	for idx, ddl := range item.DDLs {
+		prevTableStmt, postTableStmt := g.getTableForOneDDL(item, idx)
+		schemaChanged, conflictStage := g.handleDDL(item.SourceTable, prevTableStmt, postTableStmt)
+
+		switch conflictStage {
+		case optimism.ConflictDetected:
+			return nil, optimism.ConflictDetected, errors.Errorf("conflict detected for table %v", item.SourceTable)
+		case optimism.ConflictSkipWaitRedirect:
+			return newDDLs, optimism.ConflictSkipWaitRedirect, nil
+		case optimism.ConflictNone:
+			if col, err := g.checkAddDroppedColumn(ctx, item.SourceTable, ddl, prevTableStmt, postTableStmt, dropCols); err != nil {
+				return nil, optimism.ConflictError, err
+			} else if len(col) != 0 {
+				dropCols = append(dropCols, col)
+			}
+		case optimism.ConflictResolved:
+		}
+		if schemaChanged {
+			newDDLs = append(newDDLs, ddl)
+		}
+	}
+
+	if len(dropCols) > 0 {
+		if err := g.droppedColumnsStore.AddDroppedColumns(ctx, dropCols, item.SourceTable); err != nil {
+			return nil, optimism.ConflictError, err
+		}
+	}
+
+	return newDDLs, optimism.ConflictNone, nil
+}
+
+func (g *shardGroup) handleDDL(sourceTable metadata.SourceTable, prevTableStmt, postTableStmt string) (bool, optimism.ConflictStage) {
+	delete(g.conflictTables, sourceTable)
+
+	prevTable := genCmpTable(prevTableStmt)
+	postTable := genCmpTable(postTableStmt)
+
+	currTable := genCmpTable(g.normalTables[sourceTable])
+	// handle idempotent ddl
+	idempotent := false
+	if cmp, err := prevTable.Compare(currTable); err != nil || cmp != 0 {
+		if cmp, err := postTable.Compare(currTable); err == nil && cmp == 0 {
+			idempotent = true
+		}
+		log.L().Warn("prev-table not equal table saved in master", zap.Stringer("master-table", currTable), zap.Stringer("prev-table", prevTable))
+		g.normalTables[sourceTable] = prevTableStmt
```

Comment:
the definition of normalTables is table structure at checkpoint? why overwrite it to an old value